cli/medperf/logging/utils.py

```diff
@@ -0,0 +1,274 @@
+def filter_var_dict_for_yaml(unfiltered_dict):
+    valid_types = (str, dict, list, int, float)
+    filtered_dict = {}
+    for key, value in unfiltered_dict.items():
+        if not isinstance(value, valid_types) and value is not None:
+            try:
+                value = str(value)
+            except Exception:
+                value = "<OBJECT>"
+        if isinstance(value, dict):
+            value = filter_var_dict_for_yaml(value)
+        filtered_dict[key] = value
+
+    return filtered_dict
+
+
+def get_storage_contents():
+    storage_paths = config.storage.copy()
+    storage_paths["credentials_folder"] = {
+        "base": os.path.dirname(config.creds_folder),
+        "name": os.path.basename(config.creds_folder),
+    }
+    ignore_paths = {"datasets_folder", "predictions_folder", "results_folder"}
+    contents = {}
+
+    for pathname, path in storage_paths.items():
+        if pathname in ignore_paths:
+            contents[pathname] = "<REDACTED>"
+            continue
+        full_path = os.path.join(path["base"], path["name"])
+        p = subprocess.Popen(
+            ["ls", "-lR", full_path],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+        output, _ = p.communicate()
+        if p.returncode != 0:
+            contents[pathname] = "Could not retrieve tree for storage paths"
+        contents[pathname] = output
+
+    return contents
+
+
+def get_installed_packages():
+    installed_packages = {}
+    for package in pkg_resources.working_set:
+        installed_packages[package.key] = package.version
+    return installed_packages
+
+
+def get_python_environment_information():
+    environment_info = {
+        "Operating System": platform.system(),
+        "Operating System Release": platform.release(),
+        "Operating System Version": platform.version(),
+        "Python Implementation": platform.python_implementation(),
+        "Python Version": platform.python_version(),
+        "Python Compiler": platform.python_compiler(),
+        "Python Build": platform.python_build(),
+        "Machine Architecture": platform.machine(),
+        "Processor Type": platform.processor(),
+        "Python Executable": shutil.which("python"),
+        "Installed Modules": get_installed_packages(),
+    }
+    return environment_info
+
+
+def get_gpu_information():
+    try:
+        gpu_info = {}
+        # Get GPU information
+        p = subprocess.Popen(
+            [
+                "nvidia-smi",
+                "--query-gpu=index,gpu_name,driver_version,compute_cap,memory.total",
+                "--format=csv",
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+        output, _ = p.communicate()
+        if p.returncode != 0:
+            raise ExecutionError("nvidia-smi not installed/available")
+        gpus_data = [row for row in csv.DictReader(output.split("\n"))]
```

Comment:
what if the output contained a warning message? this may fail right

Reply:
I'm now using nvidia-smi's -f parameter to store the query in a csv file and I'm parsing that file then. Would that work, or would you prefer just dumping the contents?